### PR TITLE
Relationship labels, Label self deployed resources

### PR DESCRIPTION
**Pull request by oshoval**
In order to have relationship (app.kubernetes.io/*)
on all deployed resources,
this commit handles the resources that are deployed by itself
such as:
1. Node Network Configuration Enactments
2. Node Network State

The labels are inherited from the pod, if exists.
The pod will have those labels if deployed by OLM / HCO,
otherwise it won't and the labels won't be added to the above
resources.

Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

### CODE REVIEW

**Comment by oshoval** [comment]
@RamLavi 
need please ok-to-test

trying to figure out how to add tests, since e2e are without HCO/OLM
and for unit tests it will need mocking the os.env

File: pkg/names/names.go [hunk 1]
```diff
@@ -1,7 +1,36 @@
...
+
+	for key, label := range mapLabelKeys {
+		envVar := os.Getenv(key)
+		if envVar != "" {
+			labels[label] = envVar
+		}
+	}
+
+	return labels
+}
```
Comment (qinqon) [inline pkg/names/names.go:36]: Doing this everything is not needed since env vars are inmutable on pods, let's just labels the map at startup and use it in the code.

**Review by qinqon — CHANGES_REQUESTED** [review]
Init the labels map at startup.

File: pkg/names/names.go (hunk 1 above)
Comment (oshoval) [reply pkg/names/names.go:36]: is it fine to do this with a function closure ?
this way it will init it just the first time the function run and on the sequential runs it will use the cached values

File: api/v1alpha1/nodenetworkconfigurationenactment_types.go [hunk 2]
```diff
@@ -31,16 +32,16 @@ type NodeNetworkConfigurationEnactment struct {
 }
 
 func NewEnactment(nodeName string, policy NodeNetworkConfigurationPolicy) NodeNetworkConfigurationEnactment {
+	labels := names.GetRelationshipLabels()
```
Comment (RamLavi) [inline api/v1alpha1/nodenetworkconfigurationenactment_types.go:35]: @qinqon do we want that set on v1alpha1?
i don't think it's braking anything but not sure it's necessary

**Review by RamLavi — COMMENTED** [review]
@oshoval, related to your question about unit tests, I think GetRelationshipLabelscan be unit-tested by playing with `os.Setenv()`.

**Comment by oshoval** [comment]
Addressed https://github.com/nmstate/kubernetes-nmstate/pull/770#discussion_r648042017
will work on unit tests

File: api/v1alpha1/nodenetworkconfigurationenactment_types.go [hunk 3]
```diff
@@ -31,16 +32,16 @@ type NodeNetworkConfigurationEnactment struct {
 }
 
 func NewEnactment(nodeName string, policy NodeNetworkConfigurationPolicy) NodeNetworkConfigurationEnactment {
+	labels := names.RelationshipLabels
+	labels[shared.EnactmentPolicyLabel] = policy.Name
```
Comment (oshoval) [inline api/v1alpha1/nodenetworkconfigurationenactment_types.go:36]: this might change the names.RelationshipLabels
checking

**Comment by oshoval** [comment]
Added unit tests

File: pkg/names/names.go [hunk 4]
```diff
@@ -1,7 +1,36 @@
...
+		"MANAGED_BY": MANAGED_BY_LABEL_KEY,
+	}
+
+	for key, label := range mapLabelKeys {
+		envVar := os.Getenv(key)
+		if envVar != "" {
+			RelationshipLabels[label] = envVar
+		}
+	}
+}
```
Comment (qinqon) [inline pkg/names/names.go:36]: I think I was wrong, since we are just reading env vars we have no need to do this at startup time, let's just read the env vars in this function and update a map instead of storing at a global var with is a worse solution, sorry about the back and forth.

Also since we have use cases where a label is already defined (enactments) is better to update a map instead of creating a new one `func IncludeRelationshipLabels(map[string]string) map[string]string`, so we can use it declaratively with a little functional programming.  

```golang
Labels: IncludeRelationshipLabels(map[string]string{shared.EnactmentPolicyLabel: policy.Name}), 
```

File: pkg/names/names_test.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("InitRelationshipLabels", func() {
+	Context("When env vars are empty", func() {
```
Comment (qinqon) [inline pkg/names/names_test.go:11]: We need to add unit test for env vars not present to ensure that is still working and partial empty and partial not empty.

**Review by qinqon — CHANGES_REQUESTED** [review]
Let's remove the global var, it was a bad idea from myself.

**Comment by oshoval** [comment]
Addressed comments

File: pkg/names/names_test.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+			os.Unsetenv("MANAGED_BY")
+		})
+
+		It("Should return labels with the right values", func() {
+			labels := IncludeRelationshipLabels(map[string]string{"foo": "bar"})
+			Expect(labels["foo"]).To(Equal("bar"))
+			Expect(labels[COMPONENT_LABEL_KEY]).To(Equal(expectedComponentLabel))
+			Expect(labels[PART_OF_LABEL_KEY]).To(Equal(expectedPartOfLabel))
+			Expect(labels[VERSION_LABEL_KEY]).To(Equal(""))
+			Expect(labels[MANAGED_BY_LABEL_KEY]).To(Equal(""))
```
Comment (RamLavi) [inline pkg/names/names_test.go:73]: if `MANAGED_BY_LABEL_KEY` is not set, don't we expect the output label to not include it at all?

File: pkg/names/names_test.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"os"
+	"reflect"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("IncludeRelationshipLabels", func() {
+	Context("When env vars don't exist", func() {
```
Comment (RamLavi) [inline pkg/names/names_test.go:12]: we should also include the case were the labels are already set (upgrade scenario)

**Review by RamLavi — CHANGES_REQUESTED** [review]
looks good overall, added comments

File: pkg/names/names_test.go (hunk 6 above)
Comment (oshoval) [reply pkg/names/names_test.go]: it doesn't exists, however even element that doesn't exist return `""`
i changed it to something more explicit

nice catch

**Comment by oshoval** [comment]
addressed comments

File: api/v1alpha1/nodenetworkconfigurationenactment_types.go (hunk 2 above)
Comment (oshoval) [reply api/v1alpha1/nodenetworkconfigurationenactment_types.go]: @qinqon can you see this one please ?
